File: PGBooks/PGBooks/Sources/UI/Common/Views/CustomViews.m
//
//  CustomViews.m
//  PGBooks
//
//  Created by ipagong on 13/10/2019.
//  Copyright © 2019 ipagong. All rights reserved.
//

#import "CustomViews.h"

@implementation UIView (radius)

@dynamic cornerRadius;
@dynamic borderColor;
@dynamic borderWidth;

- (void)setCornerRadius:(CGFloat)cornerRadius {
    self.layer.cornerRadius = cornerRadius;
}

- (void)setBorderColor:(UIColor *)borderColor {
    self.layer.borderColor = borderColor.CGColor;
}

- (void)setBorderWidth:(CGFloat)borderWidth {
    self.layer.borderWidth = borderWidth;
}

@end

@implementation PaddingLabel

#pragma mark - Super

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    if (self) {
        self.topInset       = 0;
        self.bottomInset    = 0;
        self.leftInset      = 0;
        self.rightInset     = 0;
    }
    return self;
}

- (void)drawTextInRect:(CGRect)rect {
    UIEdgeInsets insets = UIEdgeInsetsMake(self.topInset,
                                           self.leftInset,
                                           self.bottomInset,
                                           self.rightInset);
    [super drawTextInRect:UIEdgeInsetsInsetRect(rect, insets)];
}

- (CGSize)intrinsicContentSize {
    CGSize size = [super intrinsicContentSize];
    return CGSizeMake(size.width + self.leftInset + self.rightInset,
                      size.height + self.topInset + self.bottomInset);
}

@end

//@implementation RoundView
//- (void)setCornerRadius:(CGFloat)cornerRadius {
//    self.layer.cornerRadius = cornerRadius;
//}
//
//- (void)setBorderColor:(UIColor *)borderColor {
//    self.layer.borderColor = borderColor.CGColor;
//}
//
//- (void)setBorderWidth:(CGFloat)borderWidth {
//    self.layer.borderWidth = borderWidth;
//}
//@end

//@implementation RoundLabel
//- (void)setCornerRadius:(CGFloat)cornerRadius {
//    self.layer.cornerRadius = cornerRadius;
//}
//
//- (void)setBorderColor:(UIColor *)borderColor {
//    self.layer.borderColor = borderColor.CGColor;
//}
//
//- (void)setBorderWidth:(CGFloat)borderWidth {
//    self.layer.borderWidth = borderWidth;
//}
//@end
//
//@implementation RoundButton
//- (void)setCornerRadius:(CGFloat)cornerRadius {
//    self.layer.cornerRadius = cornerRadius;
//}
//
//- (void)setBorderColor:(UIColor *)borderColor {
//    self.layer.borderColor = borderColor.CGColor;
//}
//
//- (void)setBorderWidth:(CGFloat)borderWidth {
//    self.layer.borderWidth = borderWidth;
//}
//@end

File: PGBooks-swift/PGBooks-swift/Sources/UI/Common/Views/CustomViews.swift
//
//  CustomViews.swift
//  PGBooks-swift
//
//  Created by ipagong on 16/10/2019.
//  Copyright © 2019 ipagong. All rights reserved.
//

import UIKit

extension UIView {
    @IBInspectable
    public var cornerRadius: CGFloat {
        set { self.layer.cornerRadius = newValue }
        get { return self.layer.cornerRadius     }
    }
    
    @IBInspectable
    public var borderWidth: CGFloat {
        set { self.layer.borderWidth = newValue }
        get { return self.layer.borderWidth     }
    }
    
    @IBInspectable
    public var borderColor:UIColor? {
        set { self.layer.borderColor = newValue?.cgColor }
        get {
            guard let color = self.layer.borderColor else { return nil }
            return UIColor(cgColor: color)
        }
    }
}

class PaddingLabel : UILabel {
    @IBInspectable public var topInset      : CGFloat = 0
    @IBInspectable public var bottomInset   : CGFloat = 0
    @IBInspectable public var leftInset     : CGFloat = 0
    @IBInspectable public var rightInset    : CGFloat = 0

    override func drawText(in rect: CGRect) {
        let insets = UIEdgeInsets(top: self.topInset,
                                  left: self.leftInset,
                                  bottom: self.bottomInset,
                                  right: self.rightInset)
        
        return super.drawText(in: rect.inset(by: insets))
    }
    
    override var intrinsicContentSize: CGSize {
        let size = super.intrinsicContentSize
        
        return CGSize(width: size.width + self.leftInset + self.rightInset,
                      height: size.height + self.topInset + self.bottomInset)
    }
}
